src/storage/src/wal.rs

```diff
@@ -0,0 +1,385 @@
+
+                if !file_name.ends_with("final") {
+                    if unfinished_file.is_some() {
+                        bail!("found more than one in-progress log segments when trying to resume wal for relation {}: {} and {}",
+                            id,
+                            file_name,
+                            unfinished_file.unwrap().0.display())
+                    }
+                    let parts: Vec<_> = file_name.split('-').collect();
+                    let sequence_number = parts[1].parse::<usize>()?;
+                    unfinished_file = Some((path.to_path_buf(), sequence_number));
+                }
+            }
+        }
+
+        if let Some((unfinished_file, sequence_number)) = unfinished_file {
+            // Lets start writing to the previously created file.
+            let file = fs::OpenOptions::new()
+                .append(true)
+                .open(&unfinished_file)
+                .with_context(|| {
+                    format!(
+                        "trying to reopen wal segment file at path: {}",
+                        unfinished_file.display()
+                    )
+                })?;
+
+            let ret = Self {
+                id,
+                base_path,
+                current_file: file,
+                current_sequence_number: sequence_number,
+                current_bytes_written: 0,
+                total_bytes_written: 0,
+            };
+
+            Ok(ret)
+        } else {
+            bail!("couldn't find any in-progress log segments when trying to resume wal")
+        }
+    }
+}
+
+pub struct WriteAheadLogs {
+    path: PathBuf,
+    wals: HashMap<GlobalId, WriteAheadLog>,
+}
+
+impl WriteAheadLogs {
+    pub fn new(path: PathBuf) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            path,
+            wals: HashMap::new(),
+        })
+    }
+
+    // TODO make this path use cluster_id
+    fn get_path(&self, id: GlobalId) -> PathBuf {
+        self.path.join(id.to_string())
+    }
+
+    pub fn create(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if self.wals.contains_key(&id) {
+            bail!(
+                "asked to create wal for relation: {:?} that was already created",
+                id
+            )
+        }
+
+        let wal_base_path = self.get_path(id);
+        let wal = WriteAheadLog::create(id, wal_base_path)?;
+        self.wals.insert(id, wal);
+        Ok(())
+    }
+
+    pub fn destroy(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!(
+                "asked to delete wal for relation: {:?} that doesn't exist",
+                id
+            )
+        }
+
+        self.wals.remove(&id);
+        Ok(())
+    }
+
+    pub fn write(&mut self, id: GlobalId, updates: &[Update]) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!("asked to write to unknown relation: {}", id)
```

Comment:
This happens even in the absence of any DROP:

1. Start MZ with ```--experimental --persistent-tables```
2. Insert rows
3. Restart MZ
4. Insert some more rows

Reply:
ok I think I know why this is and I think its the same reason as the other comment where tables are visible for reads before being fully loaded up - let me investigate this a bit later today.

Reply:
I was able to reproduce and fix this specific problem locally - this was a silly oversight on my part thank you for catching it! Let me know if it still shows up for you

Reply:
This is now fixed.